<a href="https://colab.research.google.com/github/vinajuliet/python_for_data_scients/blob/main/Salinan_dari_Function.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**@create By : Dede Brahma Arianto S.Kom., M.Kom**

# Function

In [ ]:
#@title fungsi sederhana

def coba_fungsi(): # defenisi nama fungsi
	print("ini adalah fungsi") # proses fungsi


In [ ]:
coba_fungsi() #panggil nama fungsi

ini adalah fungsi


In [ ]:
#@title fungsi dengan return

def coba_fungsi2():
	var_tes = "ini adalah fungsi"
	return var_tes


In [ ]:
hasil = coba_fungsi2()
print("Hasil fungsi adalah: ", hasil)

# print("Hasil fungsi adalah: ", coba_fungsi2())

Hasil fungsi adalah:  ini adalah fungsi


In [ ]:
#@title fungsi dengan 2 return

def coba_fungsi3():
  var_1 = "ini isi variabel 1"
  var_2 = "ini isi variabel 2"
  return var_1, var_2

hasil1, hasil2 = coba_fungsi3()
print("Hasil 1 : ", hasil1)
print("Hasil 2 : ", hasil2)

# hasil1 = coba_fungsi3()
# print("Hasil 1 : ", hasil1)

Hasil 1 :  ini isi variabel 1
Hasil 2 :  ini isi variabel 2


In [ ]:
#@title fungsi dengan argument

def coba_fungsi4(a, b):
  var_a = a
  var_b = b
  var_c = var_a + var_b
  return var_c

In [ ]:
nilai_1 = 5
nilai_2 = 10

hasil_nilai = coba_fungsi4(nilai_1, nilai_2)
print("Hasil penjumlahan: ", hasil_nilai)

Hasil penjumlahan:  15


In [ ]:
hasil_nilai = coba_fungsi4(20, 5)
print("Hasil penjumlahan: ", hasil_nilai)

Hasil penjumlahan:  25


In [ ]:
hasil_nilai = coba_fungsi4(input("masukkan nilai 1 = "), input("masukkan nilai 2 = "))
print("Hasil penjumlahan: ", hasil_nilai)

masukkan nilai 1 = 10
masukkan nilai 2 = 3
Hasil penjumlahan:  103


In [ ]:
#@title fungsi dengan 2 argument dan 2 return

def coba_fungsi5(a, b):
  var_a = a * 2
  var_b = b * 4
  return var_a, var_b

hasil1, hasil2 = coba_fungsi5(10, 3)
print("Hasil 1 : ", hasil1)
print("Hasil 2 : ", hasil2)

Hasil 1 :  20
Hasil 2 :  12


### Fungsi dengan *args

Untuk args menggunakan satu tanda bintang "*" Dapat diartikan bahwa def function(*args) akan mengambil seluruh nilai dalam input mengikuti posisi yang sudah ditentukan sebagai variabelnya.

In [ ]:
def coba_fungsi6(*args):
  print(args)

In [ ]:
coba_fungsi6("apple", "banana", "cherry", "semangka")

('apple', 'banana', 'cherry', 'semangka')


In [ ]:
def coba_fungsi7(*args):
  for y in args:
    x = y
    print(x)


coba_fungsi7("apple", "banana", "cherry")

apple
banana
cherry


In [ ]:
fruits = ["apple", "banana", "cherry"]

def coba_fungsi8(*buah):
  for item in buah:
    for i in item:
      print(i)


coba_fungsi8(fruits)

apple
banana
cherry


### Fungsi dengan **kwargs

Untuk kargs menggunakan satu tanda bintang "**" Dapat diartikan bahwa def function(**kwargs) akan mengambil seluruh nilai dalam key dan valuenya yang ada dalam variabel.

In [ ]:
def coba_fungsi9(**kwargs):
  print(kwargs)

In [ ]:
coba_fungsi9(item1 = "apple", item2 = "banana", item3 = "cherry")

{'item1': 'apple', 'item2': 'banana', 'item3': 'cherry'}


In [ ]:
def coba_fungsi10(**kwargs):
  for key in kwargs:
    print(key + ": " + kwargs[key])


coba_fungsi10(item1 = "apple", item2 = "banana", item3 = "cherry")

item1: apple
item2: banana
item3: cherry


In [ ]:
mahasiswa = {"nama": "Dede", "jurusan": "IT", "semester": "5"}

def coba_fungsi11(**data_mhs):
  for key in data_mhs:
    print(key + ": " + data_mhs[key])


coba_fungsi11(**mahasiswa)

nama: Dede
jurusan: IT
semester: 5


## Nested Function (Closure)

In [ ]:
def hitung_rata_rata(nilai1, nilai2, nilai3):

  def update_nilai():
    update1 = nilai1 + 2
    update2 = nilai2 + 3
    update3 = nilai3 + 4
    return update1, update2, update3

  hasil = update_nilai()
  return hasil


In [ ]:
a = 5
b = 10
c = 15

rata_rata = hitung_rata_rata(a, b, c)
print("Hasil rata-rata: ", rata_rata)

Hasil rata-rata:  (7, 13, 19)


## Fungsi Default Parameter

In [ ]:
def coba_fungsi12(nama, umur=20):
  print("Nama: ", nama)
  print("Umur: ", umur)

coba_fungsi12("Dede")

Nama:  Dede
Umur:  20


## Fungsi Kombinasi Arguments

In [ ]:
def coba_fungsi13(message, *params,  **others):
    print(f"message: {message}")
    print(f"params: {params}")
    print(f"others: {others}")

coba_fungsi13("hello world", 1, True, ("yesn't", "nope"), say_something="how are you", name="nokia 3310", discontinued=True, year_released=2000)

message: hello world
params: (1, True, ("yesn't", 'nope'))
others: {'say_something': 'how are you', 'name': 'nokia 3310', 'discontinued': True, 'year_released': 2000}


## Fungsi dengan Matrix

In [ ]:
def coba_fungsi14(matrix=[]):
    if len(matrix) == 0:
        print("[]")

    for item in matrix:
        print(item)


In [ ]:
print("Output matrix 1:")
coba_fungsi14()

Output matrix 1:
[]


In [ ]:
print("Output matrix 2:")
coba_fungsi14([
    [1, 2],
    [5, 6],
])

Output matrix 2:
[1, 2]
[5, 6]


## Fungsi dengan Agregator

In [ ]:
def aggregate(perintah, nilai, action):
    res = action(nilai)
    print("Hasil agregate", perintah, ":", res)

In [ ]:
aggregate("total", [24, 67, 22, 98, 3, 50], sum)

Hasil agregate total : 264


In [ ]:
aggregate("Max", [24, 67, 22, 98, 3, 50], max)

Hasil agregate Max : 98


In [ ]:
aggregate("Min", [24, 67, 22, 98, 3, 50], min)

Hasil agregate Min : 3


## Soal 1

Buatlah script “kalkulator” yang terdiri dari dua parameter “x” dan “y” yang akan diisi angka dan parameter “operasi” yang akan diisi nama operasi seperti “penjumlahan”, “pengurangan”, “perkalian” dan “pembagian”

Input:
kalkulator(a = 10, b = 4, operasi = “penjumlahan”)

Output:
14


## Soal 2

Buat lah 3 fungsi yang didalamnya terdapat :

* conditional if
* condition loop
* args dan kwargs

## Soal 3

Fungsi dengan argument kenapa hasilnya 103?

In [ ]:
def kalkulator(x, y, operasi):
    if operasi == "penjumlahan":
        return x + y
    elif operasi == "pengurangan":
        return x - y
    elif operasi == "perkalian":
        return x * y
    elif operasi == "pembagian":
        if y == 0:
            return "Error: Pembagian dengan nol tidak diperbolehkan"
        return x / y
    else:
        return "Operasi tidak dikenali"



In [ ]:
hasil = kalkulator(10, 4, "penjumlahan")
print(hasil)

14
